scripts/install_backend_python_libs.py

```diff
@@ -0,0 +1,276 @@
+    information about currently installed libraries and will cause this
+    installation script to behave incorrectly.
+    """
+    # The possible strings that a metadata directory or file name can
+    # start with.
+    possible_filename_start_strings = [
+        '%s-%s' % (library, version),
+        # Some metadata folders replace the hyphens of the library name with
+        # underscores.
+        '%s-%s' % (library.replace('-', '_'), version)
+    ]
+    for f in os.listdir(common.THIRD_PARTY_PYTHON_LIBS_DIR):
+        if (f.startswith(possible_filename_start_strings[0]) or
+                f.startswith(possible_filename_start_strings[1])):
+            to_delete_path = os.path.join(
+                common.THIRD_PARTY_PYTHON_LIBS_DIR, f)
+            if os.path.isdir(to_delete_path):
+                shutil.rmtree(to_delete_path)
+            else:
+                os.remove(to_delete_path)
+
+
+def _rectify_third_party_directory(mismatches):
+    """Rectifies the 'third_party/python_libs' directory state to reflect the
+    current 'requirements.txt' file requirements. It takes a list of mismatches
+    and corrects those mismatches by installing or uninstalling packages.
+
+    Args:
+        mismatches: dict(string, tuple(string|None, string|None)). Dictionary
+            with the library names as keys and a tuple as values. The 1st
+            element of the tuple is the version string of the library required
+            by the requirements.txt file while the 2nd element is the version
+            string of the library currently installed in the
+            'third_party/python_libs' directory. If the library doesn't exist,
+            the tuple element will be None. For example, this dictionary
+            signifies that 'requirements.txt' requires flask with version 1.0.1
+            while the 'third_party/python_libs' directory contains flask 1.1.1:
+                {
+                  flask: ('1.0.1', '1.1.1')
+                }
+    """
+    # Handling 5 or more mismatches requires 5 or more individual `pip install`
+    # commands which is slower than just reinstalling all of the libraries using
+    # `pip install -r requirements.txt`.
+    if len(mismatches) >= 5:
+        if os.path.isdir(common.THIRD_PARTY_PYTHON_LIBS_DIR):
+            shutil.rmtree(common.THIRD_PARTY_PYTHON_LIBS_DIR)
+        subprocess.check_call([
+            'pip', 'install', '--target',
+            common.THIRD_PARTY_PYTHON_LIBS_DIR,
+            '--no-dependencies', '-r',
+            common.REQUIREMENTS_FILE_PATH
+        ])
+        return
+
+    for library, versions in mismatches.items():
+        requirements_version = (
+            pkg_resources.parse_version(versions[0]) if versions[0] else None)
+        directory_version = (
+            pkg_resources.parse_version(versions[1]) if versions[1] else None)
+
+        # The library is installed in the directory but is not listed in
+        # requirements.
```

Comment:
Should probably explain why you're wiping the whole dir.

Also, in this case, why continue going through the for loop? Maybe combine with the general condition above (before the for loop)?